File: src/main/java/com/jore/epoc/bo/DemandCurve.java
package com.jore.epoc.bo;

import java.math.BigDecimal;
import java.math.MathContext;
import java.math.RoundingMode;
import java.util.Objects;

import com.jore.Assert;
import com.jore.datatypes.money.Money;
import com.jore.datatypes.percent.Percent;

// TODO do not take amount out of money!
public class DemandCurve {
    public static DemandCurve create(String currency, double higherPrice, String higherPricePercent, double lowerPrice, String lowerPricePercent) {
        return new DemandCurve(Money.of(currency, higherPrice), Percent.of(higherPricePercent), Money.of(currency, lowerPrice), Percent.of(lowerPricePercent));
    }

    private final Percent lowerPricePercent;
    private final Money lowerPrice;
    private final Percent higherPricePercent;
    private final Money higherPrice;

    public DemandCurve(Money higherPrice, Percent higherPricePercent, Money lowerPrice, Percent lowerPricePercent) {
        super();
        this.lowerPricePercent = Objects.requireNonNull(lowerPricePercent, "Lower price percent must not be null");
        this.lowerPrice = Objects.requireNonNull(lowerPrice, "Lower price must not be null");
        this.higherPricePercent = Objects.requireNonNull(higherPricePercent, "Higher price percent must not be null");
        this.higherPrice = Objects.requireNonNull(higherPrice, "Higher price must not be null");
        Assert.isTrue("Lower price must be lower than higher price", lowerPrice.compareTo(higherPrice) < 0);
    }

    public Percent getDemandForPrice(Money price) {
        final Money priceDifference = higherPrice.subtract(lowerPrice);
        final Percent percentDifference = higherPricePercent.subtract(lowerPricePercent);
        final BigDecimal aValue = percentDifference.getFactorAmount().divide(priceDifference.getAmount(), new MathContext(Percent.SCALE, RoundingMode.HALF_UP));
        final BigDecimal bValue = lowerPricePercent.getFactorAmount().subtract(aValue.multiply(lowerPrice.getAmount()));
        return Percent.of(bValue.min(Objects.requireNonNull(price, "Price must not be null").getAmount().multiply(aValue).add(bValue).max(BigDecimal.ZERO)));
    }
}

File: EpocServer/Sources/App/Models/DemandCurve.swift
import Foundation

/// Linear demand curve between two price-demand points.
/// Equivalent to `com.jore.epoc.bo.DemandCurve`.
///
/// Given a higher price (low demand) and lower price (high demand), calculates
/// the percentage of the market willing to buy at any given price point.
public struct DemandCurve: Sendable {
    private let higherPrice: Money
    private let higherPricePercent: Percent
    private let lowerPrice: Money
    private let lowerPricePercent: Percent

    public init(higherPrice: Money, higherPricePercent: Percent, lowerPrice: Money, lowerPricePercent: Percent) {
        precondition(lowerPrice < higherPrice, "Lower price must be lower than higher price")
        self.higherPrice = higherPrice
        self.higherPricePercent = higherPricePercent
        self.lowerPrice = lowerPrice
        self.lowerPricePercent = lowerPricePercent
    }

    public static func create(_ currency: String, _ higherPrice: Double, _ higherPricePercent: String,
                               _ lowerPrice: Double, _ lowerPricePercent: String) -> DemandCurve {
        DemandCurve(
            higherPrice: Money.of(currency, higherPrice),
            higherPricePercent: Percent.of(higherPricePercent),
            lowerPrice: Money.of(currency, lowerPrice),
            lowerPricePercent: Percent.of(lowerPricePercent)
        )
    }

    /// Calculates the percentage of the market willing to buy at the given price.
    /// Uses linear interpolation: demand% = a * price + b
    /// Result is clamped between 0% and the maximum (b value).
    public func getDemandForPrice(_ price: Money) -> Percent {
        let priceDifference = higherPrice.subtract(lowerPrice)
        let percentDifference = higherPricePercent.subtract(lowerPricePercent)
        // a = percentDifference / priceDifference
        let aValue = percentDifference.factorAmount.doubleValue / priceDifference.amount.doubleValue
        // b = lowerPercent - a * lowerPrice
        let bValue = lowerPricePercent.factorAmount.doubleValue - aValue * lowerPrice.amount.doubleValue
        // demand = a * price + b, clamped to [0, b]
        let demand = aValue * price.amount.doubleValue + bValue
        let clamped = min(bValue, max(0, demand))
        // Round to Percent scale (6 decimal places)
        let rounded = (clamped * 1_000_000).rounded() / 1_000_000
        return Percent.of(rounded)
    }
}
